sdk/core/Azure.Core/src/DiagnosticsOptions.cs

```diff
@@ -153,5 +209,28 @@ internal DiagnosticsOptions(DiagnosticsOptions? diagnosticsOptions)
 
             return null;
         }
+
+        internal void Freeze()
+        {
+            _frozen = true;
+
+            if (LoggedHeaderNames is List<string> mutableHeaderNames)
+            {
+                LoggedHeaderNames = mutableHeaderNames.AsReadOnly();
```

Comment:
There's some complexity here in that we need to make mutable lists read-only to freeze this nested options type.